### PR TITLE
Nan rewrite

### PR DESCRIPTION
**Pull request by mmitkevich**
migrated to NAN

### CODE REVIEW

File: src/addon.cc [hunk 1]
```diff
@@ -34,46 +34,50 @@ void SetupLog(const FunctionCallbackInfo<Value>& args) {
...
-        isolate->ThrowException(Exception::TypeError(String::NewFromUtf8(isolate, "Filename argument is expected")));
+    if (!info[1]->IsString()) {
+        Nan::ThrowError("Filename argument is expected");
         return;
     }
 
-    auto filename = args[1]->ToString(isolate->GetCurrentContext()).ToLocalChecked();
-
-    mlog_configure(*String::Utf8Value(filename), false);
+    //auto filename = info[1]->ToString(); // .ToLocalChecked()
```
Comment (naughtyfox) [inline src/addon.cc]: delete this comment

File: src/addon.cc [hunk 2]
```diff
@@ -34,46 +34,50 @@ void SetupLog(const FunctionCallbackInfo<Value>& args) {
...
     mlog_set_log(configureString);
 }
 
-void Init(Local<Object> exports) {
-    Wallet::Init(exports->GetIsolate());
-    PendingTransaction::Init(exports->GetIsolate());
+
+
+NAN_MODULE_INIT(Init) {
+    //LOG("NAN 2018-09-04");
```
Comment (naughtyfox) [inline src/addon.cc]: delete this comment

File: src/addon.cc [hunk 3]
```diff
@@ -34,46 +34,50 @@ void SetupLog(const FunctionCallbackInfo<Value>& args) {
...
+    Nan::SetMethod(target, "paymentIdValid", Wallet::PaymentIdValid);
+    Nan::SetMethod(target, "addressValid", Wallet::AddressValid);
+    Nan::SetMethod(target, "setupLog", SetupLog);
 }
 
+
 NODE_MODULE(NODE_GYP_MODULE_NAME, Init)
 
-} //namespace exawallet
+} // namespace exawallet
```
Comment (naughtyfox) [inline src/addon.cc]: new line

File: example/index.js [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
-var path = 'test-wallet'
+var path = 'stagenet'
 var args = {
 	'path': path,
 	'password': '123', 
-	'network': 'mainnet',
-	'daemonAddress': 'monero.exan.tech:18081'
-	,'restoreHeight': 1608000
+	'network': 'stagenet',
+	'daemonAddress': 'Cryptonode4.prod.zorg.sh:28081'
```
Comment (naughtyfox) [inline example/index.js]: wrong stagenet url. it should be 'monero-stage.exan.tech:38081'

File: example/test-create-tx.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+var wallet;
+var sent = false;
+
+var path = 'stagenet'
+var args = {
+	'path': path,
+	'password': '123', 
+	'network': 'stagenet',
+	'daemonAddress': 'Cryptonode4.prod.zorg.sh:28081'
```
Comment (naughtyfox) [inline example/test-create-tx.js]: the same

File: example/test-create-tx.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+			console.log("wallet refreshed");
+
+			wallet.store()
+				.then(() => {console.log("wallet stored")})
+				.catch((e) => {console.log("unable to store wallet: " + e)})
+
+			if (!sent) {
+				sent = true;
+				wallet.createTransaction({
+                    'address': //mainnet:'44zrUGhyRHYbHYrfiGAtLdJMHfe5DtoFTBeVPCE6MGKzZA2bJ4tCJFuhYk3Wjp3YxEWoQU8So5xUiiArgnkBHZgX8Fyhv6e',
```
Comment (naughtyfox) [inline example/test-create-tx.js]: delete this comment

File: example/test-msig.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+var nwallets = 2
+var recreate = false;
+var args = {
+	'path': 'stagenet',
+	'password': '123', 
+	'network': 'stagenet',
+	'daemonAddress': 'Cryptonode4.prod.zorg.sh:28081'
+	,'restoreHeight': 146046
+
+	//,'mnemonic' : 'nifty inflamed against focus gasp ethics spying gulp tiger cogs evicted cohesive woken nylon erosion tell saved fatal alkaline acquire lemon maps hull imitate saved'
```
Comment (naughtyfox) [inline example/test-msig.js]: is it necessary or not? leave it or delete it

File: example/test-msig.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	reportOnWallets([sourceWallet]);
+	const onWallet = (wallet, what) => new Promise((resolve, reject) => {
+		console.log("subscribing on "+what+" for wallet "+wallet.address())
+		wallet.on(what,  () => {
+			console.log("wallet event: "+what+" for wallet "+wallet.address())
+			resolve(true)
+		})})
+	printWallets([sourceWallet])
+	console.log("waiting until refreshed")
+	//setInterval(() => printWallets([sourceWallet]),10000)
```
Comment (naughtyfox) [inline example/test-msig.js]: no commented out code please!

File: package.json [hunk 9]
```diff
@@ -31,7 +31,12 @@
     "node": "8 - 9"
   },
   "dependencies": {
-    "node-pre-gyp": "^0.9.1"
+    "del": "^3.0.0",
+    "glob": "^7.1.3",
+    "glob-promise": "^3.4.0",
+    "nan": "^2.10.0",
+    "node-pre-gyp": "^0.9.1",
+    "underscore": "^1.9.1"
```
Comment (naughtyfox) [inline package.json]: is it necessary? or it's used only in tests?

File: package.json [hunk 10]
```diff
@@ -31,7 +31,12 @@
     "node": "8 - 9"
   },
   "dependencies": {
-    "node-pre-gyp": "^0.9.1"
+    "del": "^3.0.0",
```
Comment (naughtyfox) [inline package.json]: the same

File: src/deferredtask.cc [hunk 11]
```diff
@@ -2,26 +2,28 @@
 
 namespace exawallet {
 
-DeferredTask::DeferredTask(v8::Isolate* isolate) {
-    resolver.Reset(isolate, v8::Promise::Resolver::New(isolate->GetCurrentContext()).ToLocalChecked());
+DeferredTask::DeferredTask() {
+    v8::Local<v8::Promise::Resolver> value =v8::Promise::Resolver::New(Nan::GetCurrentContext()).ToLocalChecked();
```
Comment (naughtyfox) [inline src/deferredtask.cc]: why did you rewrite it?

File: src/deferredtask.cc [hunk 12]
```diff
@@ -2,26 +2,28 @@
...
-void DeferredTask::Resolve(v8::Isolate* isolate, v8::Local<v8::Value> value) {
-    resolver.Get(isolate)->Resolve(isolate->GetCurrentContext(), value);
+void DeferredTask::Resolve(v8::Local<v8::Value> value) {
+    Nan::New(resolver)->Resolve(Nan::GetCurrentContext(), value);
 }
 
-void DeferredTask::Reject(v8::Isolate* isolate) {
-    resolver.Get(isolate)->Reject(isolate->GetCurrentContext(), v8::String::NewFromUtf8(isolate, errorString.c_str()));
+void DeferredTask::Reject() {
+    v8::Local<v8::Context> context = Nan::GetCurrentContext();
```
Comment (naughtyfox) [inline src/deferredtask.cc]: why did you rewrite it?

File: src/wallet.cc [hunk 13]
```diff
@@ -7,27 +7,29 @@
...
+
 namespace {
 
-std::string toStdString(Isolate* isolate, const Local<Value>& val) {
-  String::Utf8Value utf8val(val);
-  return std::string(*utf8val, utf8val.length());
+
+
+std::string toStdString(const Local<Value>& val) {
+    Nan::Utf8String nan_string(val);
```
Comment (naughtyfox) [inline src/wallet.cc]: why do you mix camel case and snake case? fix it!

File: src/wallet.cc [hunk 14]
```diff
@@ -59,177 +61,177 @@ bool convertNettype(Monero::NetworkType type, std::string& netstring) {
...
+        Nan::ThrowTypeError("Function accepts path to wallet");
         return;
     }
 
-    String::Utf8Value path(args[0]->ToString(isolate));
+    std::string path = toStdString(info[0]);
     auto manager = Monero::WalletManagerFactory::getWalletManager();
-    bool exists = manager->walletExists(*path);
-    args.GetReturnValue().Set(Boolean::New(isolate, exists));
+    bool exists = manager->walletExists(path); // *path
```
Comment (naughtyfox) [inline src/wallet.cc]: what is this comment for?

File: src/wallet.cc [hunk 15]
```diff
@@ -276,22 +278,22 @@ void Wallet::Init(Isolate* isolate) {
...
     tpl->InstanceTemplate()->SetInternalFieldCount(walletFunctions.size());
 
     for (const auto& info: walletFunctions) {
-        NODE_SET_PROTOTYPE_METHOD(tpl, info.name, info.func);
+        Nan::SetPrototypeMethod(tpl,info.name,info.func);
     }
-
-    constructor.Reset(isolate, tpl->GetFunction());
+    constructor.Reset(tpl->GetFunction());
+//    LOG("Wallet::Init");
```
Comment (naughtyfox) [inline src/wallet.cc]: dirty comment

File: src/wallet.h [hunk 16]
```diff
@@ -1,32 +1,36 @@
...
-#include <node.h>
-#include <node_object_wrap.h>
+#include <nan.h>
 
 #include <map>
 #include <mutex>
 
 #include <wallet2_api.h>
 
+#define LOG(x) std::cout << x << "\n";
```
Comment (naughtyfox) [inline src/wallet.h]: we already have log facilities. delete it

**Comment by naughtyfox** [comment]
delete `example/1/test-wallet*` files

File: package.json (hunk 9 above)
Comment (mmitkevich) [reply package.json]: glob & underscore are used in tests only.

File: src/deferredtask.cc (hunk 11 above)
Comment (mmitkevich) [reply src/deferredtask.cc]: There were some compilation errors I managed to fix introducing explicit type for value variable.

File: src/deferredtask.cc (hunk 12 above)
Comment (mmitkevich) [reply src/deferredtask.cc]: There were some compilation errors I managed to fix introducing explicit type for value variable.

File: example/test-create-tx.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+			console.log("wallet refreshed");
+
+			wallet.store()
+				.then(() => {console.log("wallet stored")})
+				.catch((e) => {console.log("unable to store wallet: " + e)})
+
+			if (!sent) {
+				sent = true;
+				wallet.createTransaction({
+                    'address': '56twwcC7i9zGWcDR16A3DJKERPn7CTpPJcj48QuKYumBRun7gJveNfpjfcN5qjtWAYEqGqNaXR7PVT57uVqYsmvn5QM8aim', // testnet
```
Comment (naughtyfox) [inline example/test-create-tx.js]: deprecated comment - it's `stagenet` now

File: example/test-msig.js [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+monero.setupLog(3, "xmr.log");
+
+var nwallets = 2
+var recreate = false;
+var args = {
+	'path': 'stagenet',
+	'password': '123', 
+	'network': 'stagenet',
+	'daemonAddress': 'monero-stage.exan.tech:38081'
+	,'restoreHeight': 146046
```
Comment (naughtyfox) [inline example/test-msig.js]: keep commas in one style - on the same line

File: src/addon.cc [hunk 19]
```diff
@@ -1,22 +1,22 @@
+#include <nan.h>
 #include "wallet.h"
 #include "pendingtransaction.h"
 
+
```
Comment (naughtyfox) [inline src/addon.cc]: redundant empty line

File: example/test-create-tx.js [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+var wallet;
+var sent = false;
+
+var path = 'stagenet'
+var args = {
+	'path': path,
+	'password': '123', 
+	'network': 'stagenet',
+	'daemonAddress': 'monero-stage.exan.tech:38081'
+	,'restoreHeight': 1608000
```
Comment (naughtyfox) [inline example/test-create-tx.js]: same code style pls

File: example/index.js [hunk 21]
```diff
@@ -1,17 +1,21 @@
...
+var path = 'stagenet'
 var args = {
 	'path': path,
 	'password': '123', 
-	'network': 'mainnet',
-	'daemonAddress': 'monero.exan.tech:18081'
-	,'restoreHeight': 1608000
+	'network': 'stagenet',
+	'daemonAddress': 'monero-stage.exan.tech:38081'
+	,'restoreHeight': 0
```
Comment (naughtyfox) [inline example/index.js]: same code style

**Pull request merged**